code/acting/src/acting/vehicle_controller.py

```diff
@@ -0,0 +1,217 @@
+            CarlaEgoVehicleControl,
+            f'/carla/{self.role_name}/vehicle_control_cmd',
+            qos_profile=10
+        )
+        self.status_pub: Publisher = self.new_publisher(
+            Bool,
+            f"/carla/{self.role_name}/status",
+            qos_profile=QoSProfile(
+                depth=1,
+                durability=DurabilityPolicy.TRANSIENT_LOCAL)
+        )
+
+        self.emergency_pub: Publisher = self.new_publisher(
+            Bool,
+            f"/carla/{self.role_name}/emergency",
+            qos_profile=10
+        )
+
+        self.emergency_sub: Subscriber = self.new_subscription(
+            Bool,
+            f"/carla/{self.role_name}/emergency",
+            self.__emergency_break,
+            qos_profile=10)
+
+        self.velocity_sub: Subscriber = self.new_subscription(
+            CarlaSpeedometer,
+            f"/carla/{self.role_name}/Speed",
+            self.__get_velocity,
+            qos_profile=1)
+
+        self.throttle_sub: Subscriber = self.new_subscription(
+            Float32,
+            f"/carla/{self.role_name}/throttle",
+            self.__set_throttle,
+            qos_profile=1)
+
+        self.pure_pursuit_steer_sub: Subscriber = self.new_subscription(
+            Float32,
+            f"/carla/{self.role_name}/pure_pursuit_steer",
+            self.__set_pure_pursuit_steer,
+            qos_profile=1)
+
+        self.stanley_steer_sub: Subscriber = self.new_subscription(
+            Float32,
+            f"/carla/{self.role_name}/stanley_steer",
+            self.__set_stanley_steer,
+            qos_profile=1)
+
+        self.__emergency: bool = False
+        self.__throttle: float = 0
+        self.__pure_pursuit_steer: float = 0
+        self.__stanley_steer: float = 0
+
+    def run(self):
+        """
+        Starts the main loop of the node and send a status msg.
+        :return:
+        """
+        self.status_pub.publish(True)
+        self.loginfo('VehicleController node running')
+
+        def loop(timer_event=None) -> None:
+            """
+            Collects all data received and sends a CarlaEgoVehicleControl msg.
+            :param timer_event: Timer event from ROS
+            :return:
+            """
+            # self.loginfo('Vehicle Controller Running')
+
+            controller = self.__choose_controller()
+            if controller == PURE_PURSUIT_CONTROLLER:
+                self.logdebug('Using PURE_PURSUIT_CONTROLLER')
+                steer = self.__pure_pursuit_steer
+            elif controller == STANLEY_CONTROLLER:
+                self.logdebug('Using STANLEY_CONTROLLER')
+                steer = self.__stanley_steer
+            else:
+                self.logerr("Vehicle Controller couldn't find requested "
+                            "controller.")
+                raise Exception("Requested Controller not found")
+
+            message = CarlaEgoVehicleControl()
```

Comment:
This looks good, however I am not quite sure if/how you are blocking the normal `message` from getting published in an emergency. I think it might make sense to check if `self.__emergency == true` and react accordingly?

Reply:
You're right, I forgot that. Fixed in 51f692b3a0e3fe1541e96c942793f537642e8f85